cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,329 @@
+                        return ineq_op(thrust::get<1>(right_tuple), thrust::get<1>(left_tuple));
+                    });			        
+    } else {
+        auto ineq_op = row_inequality_comparator<false>(right_table, left_table, nulls_are_smallest, asc_desc.data().get()); 
+        thrust::merge(rmm::exec_policy(stream)->on(stream),
+                    left_begin_zip_iterator,
+                    left_end_zip_iterator,
+                    right_begin_zip_iterator,
+                    right_end_zip_iterator,
+                    output_zip_iterator,
+                    [=] __device__ (thrust::tuple<gdf_size_type, gdf_size_type> const & right_tuple,
+                                    thrust::tuple<gdf_size_type, gdf_size_type> const & left_tuple) {
+                        return ineq_op(thrust::get<1>(right_tuple), thrust::get<1>(left_tuple));
+                    });					        
+    }
+
+    CHECK_STREAM(stream);
+
+    return std::make_pair(out_table_indices, out_row_indices);
+}
+
+} // namespace
+
+namespace cudf {
+namespace detail {
+
+table merge(table const& left_table,
+            table const& right_table,
+            std::vector<gdf_size_type> const& key_cols,
+            std::vector<order_by_type> const& asc_desc,
+            bool nulls_are_smallest,
+            cudaStream_t stream = 0) {
+    CUDF_EXPECTS(left_table.num_columns() == right_table.num_columns(), "Mismatched number of columns");
+    if (left_table.num_columns() == 0) {
+        return cudf::empty_like(left_table);
+    }
+    
+    std::vector<gdf_dtype> left_table_dtypes = cudf::column_dtypes(left_table);
+    std::vector<gdf_dtype> right_table_dtypes = cudf::column_dtypes(right_table);
+    CUDF_EXPECTS(std::equal(left_table_dtypes.cbegin(), left_table_dtypes.cend(), right_table_dtypes.cbegin(), right_table_dtypes.cend()), "Mismatched column dtypes");
+    CUDF_EXPECTS(key_cols.size() > 0, "Empty key_cols");
+    CUDF_EXPECTS(key_cols.size() <= static_cast<size_t>(left_table.num_columns()), "Too many values in key_cols");
+    CUDF_EXPECTS(asc_desc.size() > 0, "Empty asc_desc");
+    CUDF_EXPECTS(asc_desc.size() <= static_cast<size_t>(left_table.num_columns()), "Too many values in asc_desc");
+    CUDF_EXPECTS(key_cols.size() == asc_desc.size(), "Mismatched size between key_cols and asc_desc");
+
+
+    auto gdf_col_deleter = [](gdf_column *col) {
+        gdf_column_free(col);
+        delete col;
+    };
+    using gdf_col_ptr = typename std::unique_ptr<gdf_column, decltype(gdf_col_deleter)>;
+    std::vector<gdf_col_ptr> temp_columns_to_free;
+    std::vector<gdf_column*> left_cols_sync(const_cast<gdf_column**>(left_table.begin()), const_cast<gdf_column**>(left_table.end()));
+    std::vector<gdf_column*> right_cols_sync(const_cast<gdf_column**>(right_table.begin()), const_cast<gdf_column**>(right_table.end()));
+    for (gdf_size_type i = 0; i < left_table.num_columns(); i++) {
```

Comment:
This is the kind of functionality that should be moved into a utility function. Luckily, it already exists: https://github.com/rapidsai/cudf/blob/branch-0.9/cpp/src/string/nvcategory_util.hpp#L45

Reply:
I'm using `sync_column_categories` [here](https://github.com/rapidsai/cudf/pull/2049/files#diff-ef8fa6376fd285383b7c5011ef5af737R237). Did you mean something else?

Reply:
Ah, I see, you're right. I guess it just got lost in all the extra code here that is necessary in order to call it.